File: src/main/resources/coremods/registry_load.js
function initializeCoreMod() {
    var ASMAPI = Java.type('net.neoforged.coremod.api.ASMAPI');
    var Opcodes = Java.type('org.objectweb.asm.Opcodes');
    var InsnList = Java.type('org.objectweb.asm.tree.InsnList');
    var LabelNode = Java.type('org.objectweb.asm.tree.LabelNode');
    var InsnNode = Java.type('org.objectweb.asm.tree.InsnNode');
    var JumpInsnNode = Java.type('org.objectweb.asm.tree.JumpInsnNode');
    var VarInsnNode = Java.type('org.objectweb.asm.tree.VarInsnNode');
    
    return {
        'registry_load': {
            'target': {
                'type': 'METHOD',
                'class': 'net.minecraft.server.WorldLoader',
                'methodName': 'load',
                'methodDesc': '(Lnet/minecraft/server/WorldLoader$InitConfig;Lnet/minecraft/server/WorldLoader$WorldDataSupplier;Lnet/minecraft/server/WorldLoader$ResultFactory;Ljava/util/concurrent/Executor;Ljava/util/concurrent/Executor;)Ljava/util/concurrent/CompletableFuture;'
            },
            'transformer': function (method) {
                var target = new InsnList();
                target.add(new InsnNode(Opcodes.DUP));
                target.add(ASMAPI.buildMethodCall(
                    'org/moddingx/libx/impl/libxcore/CoreRegistryLoad', 'afterWorldGenLayerLoad',
                    '(Lnet/minecraft/core/LayeredRegistryAccess;)V',
                    ASMAPI.MethodType.STATIC
                ));
                var foundWorldGenField = false;
                for (var i = 0; i < method.instructions.size(); i++) {
                    var node = method.instructions.get(i);
                    if (node.getOpcode() == Opcodes.GETSTATIC && !foundWorldGenField) {
                        var fieldNode = node;
                        if (fieldNode.owner == 'net/minecraft/server/RegistryLayer' && fieldNode.name == 'WORLDGEN') {
                            foundWorldGenField = true;
                        }
                    }
                    else if (node.getOpcode() == Opcodes.INVOKESTATIC && foundWorldGenField) {
                        var methodNode = node;
                        if (methodNode.owner == 'net/minecraft/server/WorldLoader' && methodNode.name == 'loadAndReplaceLayer' && methodNode.desc == '(Lnet/minecraft/server/packs/resources/ResourceManager;Lnet/minecraft/core/LayeredRegistryAccess;Lnet/minecraft/server/RegistryLayer;Ljava/util/List;)Lnet/minecraft/core/LayeredRegistryAccess;') {
                            method.instructions.insert(node, target);
                            return method;
                        }
                    }
                }
                throw new Error('Failed to patch WorldLoader.class');
            }
        }
    };
}

File: coremods/src/main/java/org/moddingx/libx/coremods/transformers/RegistryLoad.java
package org.moddingx.libx.coremods.transformers;

import net.neoforged.neoforgespi.transformation.ProcessorName;
import net.neoforged.neoforgespi.transformation.SimpleClassProcessor;
import net.neoforged.neoforgespi.transformation.SimpleTransformationContext;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.*;

import java.util.Set;

public class RegistryLoad extends SimpleClassProcessor {

    private static final String LAYERED_ACCESS = "net/minecraft/core/LayeredRegistryAccess";
    private static final String REPLACE_FROM = "replaceFrom";
    private static final String REPLACE_FROM_DESC = "(Ljava/lang/Object;[Lnet/minecraft/core/RegistryAccess$Frozen;)Lnet/minecraft/core/LayeredRegistryAccess;";
    private static final String REGISTRY_LAYER = "net/minecraft/server/RegistryLayer";
    private static final String WORLDGEN_FIELD = "WORLDGEN";

    @Override
    public ProcessorName name() {
        return new ProcessorName("libx", "registry_load");
    }

    @Override
    public Set<Target> targets() {
        return Set.of(new Target("net.minecraft.server.WorldLoader"));
    }

    @Override
    public void transform(ClassNode cn, SimpleTransformationContext ctx) {
        // Registry loading in WorldLoader#load is asynchronous, so the layer replacement for the worldgen
        // layer does not live in the load method itself but in one of the lambda bodies it creates.
        // Therefore every method of the class is searched for the anchor instead of only the load method.
        for (MethodNode mn : cn.methods) {
            if (RegistryLoad.patch(mn)) {
                return;
            }
        }
        throw new IllegalStateException("Failed to patch WorldLoader.class: worldgen layer replacement not found");
    }

    private static boolean patch(MethodNode mn) {
        boolean foundWorldGenField = false;
        for (int i = 0; i < mn.instructions.size(); i++) {
            AbstractInsnNode node = mn.instructions.get(i);
            if (node.getOpcode() == Opcodes.GETSTATIC && node instanceof FieldInsnNode fieldInsnNode) {
                if (fieldInsnNode.owner.equals(REGISTRY_LAYER) && fieldInsnNode.name.equals(WORLDGEN_FIELD)) {
                    foundWorldGenField = true;
                }
            } else if (node.getOpcode() == Opcodes.INVOKEVIRTUAL && foundWorldGenField && node instanceof MethodInsnNode methodInsnNode) {
                if (methodInsnNode.owner.equals(LAYERED_ACCESS)
                        && methodInsnNode.name.equals(REPLACE_FROM)
                        && methodInsnNode.desc.equals(REPLACE_FROM_DESC)) {

                    InsnList injection = new InsnList();
                    injection.add(new InsnNode(Opcodes.DUP));
                    injection.add(new MethodInsnNode(
                            Opcodes.INVOKESTATIC,
                            "org/moddingx/libx/impl/libxcore/CoreRegistryLoad",
                            "afterWorldGenLayerLoad",
                            "(Lnet/minecraft/core/LayeredRegistryAccess;)V",
                            false
                    ));
                    mn.instructions.insert(node, injection);
                    return true;
                }
            }
        }
        return false;
    }
}
